engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -178,28 +179,41 @@ class WorkflowActor(workflowId: WorkflowId,
     case Event(WorkflowInitializationFailedResponse(reason), _) =>
       context.parent ! WorkflowFailedResponse(workflowId, InitializingWorkflowState, reason)
       goto(WorkflowFailedState)
   }
 
   when(ExecutingWorkflowState) {
-    case Event(WorkflowExecutionSucceededResponse, WorkflowActorData(_, Some(workflowDescriptor))) =>
+    case Event(response: WorkflowExecutionActorResponse, data@WorkflowActorData(_, Some(workflowDescriptor), _)) =>
       val finalizationActor = context.actorOf(WorkflowFinalizationActor.props(workflowId, workflowDescriptor), name = s"WorkflowFinalizationActor-$workflowId")
       finalizationActor ! StartFinalizationCommand
-      goto(FinalizingWorkflowState) using WorkflowActorData(finalizationActor, workflowDescriptor)
-    case Event(WorkflowExecutionFailedResponse(reasons), _) =>
-      context.parent ! WorkflowFailedResponse(workflowId, ExecutingWorkflowState, reasons)
-      goto(WorkflowFailedState)
+      goto(FinalizingWorkflowState) using data.copy(currentLifecycleStateActor = Option(finalizationActor), executionResponse = Option(response))
   }
 
   when(FinalizingWorkflowState) {
-    case Event(WorkflowFinalizationSucceededResponse, stateData) =>
-      context.parent ! WorkflowSucceededResponse(workflowId)
-      goto(WorkflowSucceededState)
-    case Event(WorkflowFinalizationFailedResponse(reasons), _) =>
-      context.parent ! WorkflowFailedResponse(workflowId, FinalizingWorkflowState, reasons)
+    case Event(WorkflowFinalizationSucceededResponse, data) =>
+      data.executionResponse match {
+        case Some(WorkflowExecutionSucceededResponse) =>
+          context.parent ! WorkflowSucceededResponse(workflowId)
+          goto(WorkflowSucceededState)
+        case Some(WorkflowExecutionFailedResponse(reason)) =>
+          context.parent ! WorkflowFailedResponse(workflowId, FinalizingWorkflowState, reason)
+          goto(WorkflowFailedState) using data.copy(currentLifecycleStateActor = None)
+        case None =>
+          context.parent ! WorkflowFailedResponse(workflowId, FinalizingWorkflowState, List(new IllegalStateException(s"Workflow $workflowId at Finalize stage without execution result.")))
+          goto(WorkflowFailedState)
+      }
+    case Event(WorkflowFinalizationFailedResponse(finalizationFailures), data) =>
+      data.executionResponse match {
+        case Some(WorkflowExecutionFailedResponse(executionFailures)) =>
+          context.parent ! WorkflowFailedResponse(workflowId, FinalizingWorkflowState, executionFailures ++ finalizationFailures)
+        case Some(WorkflowExecutionSucceededResponse) =>
+          context.parent ! WorkflowFailedResponse(workflowId, FinalizingWorkflowState, finalizationFailures)
+        case None =>
+          context.parent ! WorkflowFailedResponse(workflowId, FinalizingWorkflowState, List(new IllegalStateException(s"Workflow $workflowId at Finalize stage without execution result.")))
+      }
```

Comment:
Since I didn't add the 'abort' case here... is it easy to add that to your PR after mine is merged in?